wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift

```diff
@@ -925,10 +925,25 @@ extension ZMUserSession: ZMSyncStateDelegate {
             context: notificationContext
         ).post()
 
-        let selfClient = ZMUser.selfUser(in: syncContext).selfClient()
-        if selfClient?.hasRegisteredMLSClient == true {
+        WaitingGroupTask(context: syncContext) { [self] in
+            await fetchBackendMLSPublicKeys()
+            await fetchAndStoreFeatureConfig()
+
+            let (selfClient, hasRegisteredMLSClient) = await syncContext.perform {
+                let selfClient = ZMUser.selfUser(in: self.syncContext).selfClient()
+                let hasRegisteredMLSClient = selfClient?.hasRegisteredMLSClient == true
+                return (selfClient, hasRegisteredMLSClient)
+            }
 
-            WaitingGroupTask(context: syncContext) { [self] in
+            // If we discover that
+            // the MLS feature is enabled, there are MLS public keys on the backend and there is no registered MLS client,
+            // we should create one.
+            let needsToRegisterMLSClient = BackendInfo.isMLSEnabled && mlsFeature.isEnabled && !hasRegisteredMLSClient
+            if let selfClient, needsToRegisterMLSClient {
+                await createMLSClient(client: selfClient)
```

Comment:
try to avoid passing a managed object (the self client). It seems that all that is needed are some ids.